cEP-0005.md

```diff
@@ -0,0 +1,157 @@
+
+The following changes are proposed and illustrated in the mockup following this
+paragraph:
+
+- Implicit `default` section inheritance will be removed. (It will be kept with
+  warning for some time to allow people to do the switch.)
+- Explicit inheritance is possible by giving sections namespaces (`all.python`
+  would inherit from the `all` section.)
+- Values can be added to inherited values. E.g. `ignore += ./vendor` would take
+  the inherited value and add `, ./vendor` to it. Note that additions within
+  a section will not be possible! The configuration should describe a state and
+  not a program and must remain flow insensitive.
+- To specify, what analysis should be run the user will not have to care about
+  bears. Instead, an `aspects` setting will be provided allowing the user to
+  specify what aspects of his code should be analyzed. Aspects could be
+  `spelling` or `smells` and are namespaced such that `redundancy.clones` will
+  be possible as well. Aspects can be defined inclusive (analyze those aspects,
+  with the `aspects` settings) or exclusive (analyze all but those aspects,
+  with the `ignore_aspects` setting).
+- Specifying `bears` manually will still be possible as it eases use especially
+  for bear developers.
+
+### Mockup
+
+The following is a mockup of how a configuration could look like.
+
+```ini
+[all]
+# Values can be added to inherited ones only, not within the section because
+# a configuration should describe a state and not involve operations.
+# A system wide coafile can define venv, .git, ... and we would recommend to
+# always use += for ignore even in the default section to inherit those values.
+# += always concatenates with a comma.
+ignore += ./vendor
+
+max_line_length = 80
+
+# This inherits all settings from the `all` section. There is no implicit
+# inheritance.
+[all.python]
+language = Python
+files = **.py
+aspects = smell, redundancy.clone  # Inclusive
+
+[all.c]
+language = C
+files = **.(c|h)
+ignore_aspects = redundancy  # Exclusive
+```
+
+### Bear API
+
+Bears will receive the aspects that should be checked as a parameter and
+results will get metadata to indicate what aspect has been checked.
+
+```python
+from ... import LocalBear
+from ... import Result
+from ...Aspects import Redundancy
+
+
+class RedundancyBear(LocalBear):
+    # coala will run bears based on this metadata
+    LANGUAGES = {"C", "Python"}
+    ASPECTS = {Redundancy.Clone, Redundancy.UnusedImport}
+
+    # Aspects are passed as parameter
+    def run(self, filename, file, aspects):
+        # No documentation required anymore for the bears and their parameters.
+
+        # Check the code ...
+
+        # A bear may yield results for any aspect even if it's not selected but
+        # may use the selected aspects to save performance. coala will filter
+        # out only selected aspects.
+        yield Result.from_values(aspect=Redundancy.Clone, ...)
+```
```

Comment:
what about making one-function-per-aspect? I can't think of an example currently where different aspects need to be taken into account together^^